### PR TITLE
APIv2 Apiary spec: complete regs and subs operations + updateContext + other fixes

### PR DESCRIPTION
**Pull request by fgalan**
RESTfull-based operation for manage registration and operations has been added (PATCH/GET/DELETE), along with updateContext RPC-based operation. Apart from that, the following changes/fixed have been implemented:

* RESTfull-based subscription and registration doesn't allow a list of entities.
* URL schema `http` is the only mandatory for notifications, others are taken into account but in an optional way
* `id` for exact entity ID matching, `idPattern` for entity ID patterns.
* Removing of `options` boilerplate, progressing `duration` to top level.
* Fixed `conditions` description.
* Removing `attrs` from `subject` (given that now is part of `conditions`).

### CODE REVIEW

File: doc/apiary/v2/fiware-ngsiv2-reference.apib [hunk 1]
```diff
@@ -1045,56 +1045,62 @@ A subscription is represented by a JSON object with the following fields:
...
 * `notification` : It is an object that describes the notification received by the subscriber.
-* `options` : It is a dictionary which can contain different parameters related to the subscription, including:
-  * `duration`: Duration of the subscription in ISO8601 format. Infinite if not specified.
+* `duration`: Duration of the subscription in ISO8601 format. Infinite if not specified.
 
-A 'subject' contains the following subfields:
+A `subject` contains the following subfields:
 
-* `entity` : Id or pattern id which identifies the affected entities (optional)
+* `id` : Id of the affected entities (optional). It cannot be use at the same time than `idPattern`.
```
Comment (jmcanterafonseca) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Can we abbreviate by saying that id or idPattern one of them must be present?

File: doc/apiary/v2/fiware-ngsiv2-reference.apib [hunk 2]
```diff
@@ -1045,56 +1045,62 @@ A subscription is represented by a JSON object with the following fields:
...
+* `idPattern` : Id pattern of the affected entities (optional). It cannot be use at the same time than `id`.
 * `type` : Type of the affected entities (optional)
-* `attributes`: List of attributes to be monitored (if not specified it would mean all)
-* `condition`: Condition that will trigger the notification. It could be 'ONCHANGE' or 'ONTIMEINTERVAL'.
-By default will be ONCHANGE.
-* `expression`: Logical the expression on attributes. When its evaluation results in a
-`true` value the notification will be sent.
+* `condition`: Condition that will trigger the notification. It can have two optional properties, `attributes` (an array of
```
Comment (jmcanterafonseca) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: can we describe the two subproperties described as sub-ballet points?

File: doc/apiary/v2/fiware-ngsiv2-reference.apib [hunk 3]
```diff
@@ -1045,56 +1045,62 @@ A subscription is represented by a JSON object with the following fields:
...
+attribute names) and `expression` (an expression composed of `q`, `geometry` and `coords`, see "List entities" operation above
+about this field):
+ * If `attributes` and `expression` are used, a notification is sent whenever one of the attributes in the `attributes` list changes and
+   at the same time `expression` matches.
```
Comment (jmcanterafonseca) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: nit: line too long

File: doc/apiary/v2/fiware-ngsiv2-reference.apib [hunk 4]
```diff
@@ -1045,56 +1045,62 @@ A subscription is represented by a JSON object with the following fields:
...
+ * If `attributes` is used and `expression` is not used, a notification is sent whenever one of the attributes in the `attributes` list changes-
+ * If `attributes` is not used and `expression` is used, a notification is sent whenever any of the attributes of the entity changes and
+   at the same time `expression` matches.
+ * If neither `attributes` or `expression` are used, a notification is sent whenever any of the attributes of the entity changes.
 
-Entity or type must be present.
+`id`/`idPattern` or `type` must be present.
```
Comment (jmcanterafonseca) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: please put this restriction in context (above)

File: doc/apiary/v2/fiware-ngsiv2-reference.apib [hunk 5]
```diff
@@ -1045,56 +1045,62 @@ A subscription is represented by a JSON object with the following fields:
...
+                    "id": "Bcn_Welt",
                     "type": "Room",
-                    "attributes": [
-                        "temperature"
-                    ],
-                    "condition": "ONCHANGE"
+                    "condition": {
+                       "attrs:" [ "temperature " ],
+                       "expression:" {
+                          "q": "tempature>40"
```
Comment (jmcanterafonseca) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: temperature

File: doc/apiary/v2/fiware-ngsiv2-reference.apib [hunk 6]
```diff
@@ -1103,24 +1109,29 @@ Returns a list of all the subscriptions present in the system
...
             "subject": {
                 "type": "Room",
-                "attributes": [
-                      "humidity"
-                ],
-                "condition": "ONTIMEINTERVAL"
+                "condition": {
+                    "attributes:" [ "temperature " ],
+                    "expression:" {
+                       "q": "tempature>40"
```
Comment (jmcanterafonseca) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: temperature

File: doc/apiary/v2/fiware-ngsiv2-reference.apib [hunk 7]
```diff
@@ -1129,6 +1140,73 @@ The subscription is represented by a JSON object as described at the beginning o
 
             Location: /v2/subscriptions/abcde98765
 
+### Get subscription [GET /v2/subscription/{subId}]
+
```
Comment (jmcanterafonseca) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: /v2/subscriptions/{subscriptionId}

File: doc/apiary/v2/fiware-ngsiv2-reference.apib [hunk 8]
```diff
@@ -1129,6 +1140,73 @@ The subscription is represented by a JSON object as described at the beginning o
...
++ Response 200 (application/json)
+
+        {
+            "id": "abcdef",
+            "subject": {
+                "type": "Room",
+                "condition": {
+                    "attributes:" [ "temperature " ],
+                    "expression:" {
+                       "q": "tempature>40"
```
Comment (jmcanterafonseca) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: temperature

File: doc/apiary/v2/fiware-ngsiv2-reference.apib [hunk 9]
```diff
@@ -1129,6 +1140,73 @@ The subscription is represented by a JSON object as described at the beginning o
...
+            },
+            "notification": {
+                "callback": "http://localhost:1234",
+                "attributes": ["temperature", "humidity"],
+                "throttling": "PT5S"
+            },
+            "duration": "PT1M"
+        }
+
+### Update subscription [PATCH /v2/subscription/{subId}]
```
Comment (jmcanterafonseca) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: PATCH /v2/subscriptions/{subscriptionId}

File: doc/apiary/v2/fiware-ngsiv2-reference.apib [hunk 10]
```diff
@@ -1129,6 +1140,73 @@ The subscription is represented by a JSON object as described at the beginning o
...
+
++ Request (application/json)
+
+        {
+            "duration": "PT1M"
+        }
+
++ Response 204
+
+### Delete subscription [DELETE /v2/subscription/{subId}]
```
Comment (jmcanterafonseca) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: DELETE /v2/subscriptions/{subscriptionId}

File: doc/apiary/v2/fiware-ngsiv2-reference.apib [hunk 11]
```diff
@@ -1140,21 +1218,27 @@ A context registration is represented by a JSON object with the following fields
...
 A `subject` contains the following subfields:
 
+* `id` : Id of the affected entities (optional). It cannot be use at the same time than `idPattern`.
+* `idPattern` : Id pattern of the affected entities (optional). It cannot be use at the same time than `id`.
 * `entity` : Id or pattern id which identifies the affected entities (optional).
 * `type` : Type of the affected entities (optional).
 * `attributes`: List of attributes to be provided (if not specified it would mean all).
 
-Entity or type must be present.
+`id`/`idPattern` or `type` must be present.
```
Comment (jmcanterafonseca) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: move this statement above

File: doc/apiary/v2/fiware-ngsiv2-reference.apib [hunk 12]
```diff
@@ -1201,6 +1286,65 @@ The registration is represented by a JSON object as described at the beginning o
 
             Location: /v2/registrations/abcde98765
 
+### Get context provider registration [GET /v2/registrations/{regId}]
```
Comment (jmcanterafonseca) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: /v2/registrations/{registrationId}

File: doc/apiary/v2/fiware-ngsiv2-reference.apib [hunk 13]
```diff
@@ -1201,6 +1286,65 @@ The registration is represented by a JSON object as described at the beginning o
...
+                "type": "Room",
+                "attributes": [
+                      "humidity"
+                ]
+            },
+            "callback":  "http://localhost:1234",
+            "duration": "PT1M"
+        }
+
+### Update context provider registration [PATCH /v2/registrations/{regId}]
```
Comment (jmcanterafonseca) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: /v2/registrations/{registrationId}

File: doc/apiary/v2/fiware-ngsiv2-reference.apib [hunk 14]
```diff
@@ -1201,6 +1286,65 @@ The registration is represented by a JSON object as described at the beginning o
...
++ Request (application/json)
+
+        {
+            "duration": "PT1M"
+        }
+
++ Response 204
+
+### Delete context provider registration [DELETE /v2/registrations/{regId}]
+
```
Comment (jmcanterafonseca) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: /v2/registrations/{registrationId}

File: doc/apiary/v2/fiware-ngsiv2-reference.apib [hunk 15]
```diff
@@ -1315,6 +1459,42 @@ Response code:
 
 ### updateContext [POST /v2/updateContext]
 
+The payload is an object which two properties:
```
Comment (jmcanterafonseca) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: with

File: doc/apiary/v2/fiware-ngsiv2-reference.apib [hunk 16]
```diff
@@ -1315,6 +1459,42 @@ Response code:
...
+
+* `action`, to specify the kind of updateContext action to do (see OMA NGSI specification for the semantics), either UPDATE, APPEND or DELETE
```
Comment (jmcanterafonseca) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: line too long

**Comment by jmcanterafonseca** [comment]
thanks Fermin good work. LGTM with the tiny requested changes fixed

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 1 above)
Comment (kzangeli) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Typo: same time *as*

And yes, we should somehow say that id|idPattern is mandatory, but just one at a time ...

Perhaps using (semi-mandatory*) and then the longer explanation "in the footer"?

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 4 above)
Comment (kzangeli) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: This here would be the "footer".  Perhaps a hyperlink?

File: doc/apiary/v2/fiware-ngsiv2-reference.apib [hunk 17]
```diff
@@ -1045,56 +1045,62 @@ A subscription is represented by a JSON object with the following fields:
...
-to model these parameters, thus this is likely to change.
+A `notification` object contains the following subfields:
 
-A 'notification' object contains the following subfields:
-
-* `attributes`: List of attributes to be included in the notification message.
-If not specified they will be the same as those which are monitored
-* `callback` : URL that will point to the service which will be invoked when a notification is generated
-* `throttling`: A period of time (in ISO8601 format) which must ellapse between two consecutive notifications.
+* `attributes`: List of attributes to be included in the notification message. If not specified they will be all entity attributes.
```
Comment (kzangeli) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: If not specified (or empty), all attributes are included in the notification

File: doc/apiary/v2/fiware-ngsiv2-reference.apib [hunk 18]
```diff
@@ -1045,56 +1045,62 @@ A subscription is represented by a JSON object with the following fields:
...
+* `callback` : URL that will point to the service which will be invoked when a notification is generated. A NGSIv2 compliant server must
```
Comment (kzangeli) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: URL pointing to the service ...
URL indicating the service ...

Can't really use future here - exactly *when* will it point to the service? (there are more cases of future in the doc)

File: doc/apiary/v2/fiware-ngsiv2-reference.apib [hunk 19]
```diff
@@ -1045,56 +1045,62 @@ A subscription is represented by a JSON object with the following fields:
...
+   support `http` URL schema, other schemas (e.g. schemas for web sockets) could also be supported.
+* `throttling`: A period of time (in ISO8601 format) which must ellapse between two consecutive notifications. It is optional.
```
Comment (kzangeli) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: typo: elapse

Wording suggestion: Minimal period of time (in ...

File: doc/apiary/v2/fiware-ngsiv2-reference.apib [hunk 20]
```diff
@@ -1045,56 +1045,62 @@ A subscription is represented by a JSON object with the following fields:
...
 
 ### List subscriptions [GET /v2/subscriptions]
 
 Returns a list of all the subscriptions present in the system
 
+Response:
+
+* Successful operation uses 200 OK
+* Errors use a non-200 and (optionally) an error payload. See subsection on Error Responses for more detail.
```
Comment (kzangeli) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: I think plural sounds a lot better here: "for more details".
Opinions?

File: doc/apiary/v2/fiware-ngsiv2-reference.apib [hunk 21]
```diff
@@ -1129,6 +1140,73 @@ The subscription is represented by a JSON object as described at the beginning o
...
+            "duration": "PT1M"
+        }
+
+### Update subscription [PATCH /v2/subscription/{subId}]
+
+Only the fields included in the request are updated in the subscription.
+
+Response:
+
+* Successful operation uses 204 No Content
```
Comment (kzangeli) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Is 204 better than simply a 200 and no payload?
I'm far from an expert on HTTP status codes, but this I can't say I like this very much ...

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 4 above)
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Given that id/idPattern and type are described in different bullets, this the very first place in the text where I can describe restrictions applying both...

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 11 above)
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Given that id/idPattern and type are described in different bullets, this the very first place in the text where I can describe restrictions applying both...

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 21 above)
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Not sure if 200 OK without payload is legal... Anyway, using 204 No Content (that seems have been invented for that purpose given its name :) make that more explicit.

NTC

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 20 above)
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: I like plural. I'll do an overall replacement to use it.

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 1 above)
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Fixed in 43fff5d

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 2 above)
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Fixed in 43fff5d

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 3 above)
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Fixed in 43fff5d

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 4 above)
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: NTC (at least, I cannot figure out how to improve this...)

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 5 above)
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Fixed in 43fff5d

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 6 above)
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Fixed in 43fff5d

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 7 above)
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Fixed in 43fff5d

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 8 above)
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Fixed in 43fff5d

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 9 above)
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Fixed in 43fff5d

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 10 above)
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Fixed in 43fff5d

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 11 above)
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: NTC (at least, I cannot figure out how to improve this...)

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 12 above)
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Fixed in 43fff5d

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 13 above)
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Fixed in 43fff5d

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 14 above)
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Fixed in 43fff5d

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 15 above)
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Fixed in 43fff5d

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 16 above)
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Fixed in 43fff5d

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 17 above)
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Fixed in 43fff5d

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 18 above)
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Fixed in 43fff5d

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 19 above)
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Fixed in 43fff5d

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 20 above)
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Fixed in 43fff5d

**Comment by kzangeli** [comment]
LGTM

**Pull request merged**